pkg/genericapiserver/server/serve.go

```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+type tcpConnectionChainHandler interface {
+	init(*net.TCPConn) error
+	wrap(net.Conn) (net.Conn, error)
+}
+
+type tcpConnectionChainListener struct {
+	*net.TCPListener
+	chain []tcpConnectionChainHandler
+}
+
+func (l tcpConnectionChainListener) Accept() (net.Conn, error) {
+	tc, err := l.AcceptTCP()
+	if err != nil {
+		return nil, err
+	}
+	for _, handler := range l.chain {
+		err = handler.init(tc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var conn net.Conn = tc
+	for _, handler := range l.chain {
+		conn, err = handler.wrap(conn)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return tc, nil
+}
+
 // tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
 // connections. It's used by ListenAndServe and ListenAndServeTLS so
 // dead TCP connections (e.g. closing laptop mid-download) eventually
 // go away.
 //
 // Copied from Go 1.7.2 net/http/server.go
-type tcpKeepAliveListener struct {
-	*net.TCPListener
+type tcpKeepAliveHandler struct {
+}
+
+func (ln tcpKeepAliveHandler) init(conn *net.TCPConn) error {
+	conn.SetKeepAlive(true)
+	conn.SetKeepAlivePeriod(defaultKeepAlivePeriod)
+	return nil
+}
+
+func (ln tcpKeepAliveHandler) wrap(conn net.Conn) (net.Conn, error) {
+	return conn, nil
 }
 
-func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
-	tc, err := ln.AcceptTCP()
+// ipBasedLimitListener accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
+//
+// Based on Go 1.7.4 net/netutil/listen.go
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+type ipBasedLimit struct {
+	// Map in which we map the serialized ip to the current count
+	limits map[string]int
+
+	// Current total count of the connections in existance (no localhost)
+	total int
+
+	// Lock used to ensure the limits and total are updated atomicaly
+	mutex sync.Mutex
+
+	// The maximum concurent connections an IP is allowed to have open
+	ipMax int
+
+	// The total maximum concurrent connections allowed open (Except localhost)
+	max int
+}
+
+func (i *ipBasedLimit) incrementIp(ipser string) error {
+	if i.max <= 0 {
+		return nil
+	}
+	ip := net.ParseIP(ipser)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", ipser)
+		return limitedError{ip: ipser}
+	}
+	if ip.IsLoopback() {
+		return nil
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	current := i.limits[ipser]
+	if current >= i.ipMax || i.total >= i.max {
+		return limitedError{ip: ipser}
+	}
+	i.total++
+	i.limits[ipser] = current + 1
+	return nil
+}
+
+func (i *ipBasedLimit) decrementIp(ipser string) {
+	if i.max <= 0 {
+		return
+	}
+	ip := net.ParseIP(ipser)
+	if ip == nil || ip.IsLoopback() {
+		return
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	i.total--
+	current := i.limits[ipser]
```

Comment:
Do we have a programming error if `ipset` does not exist anymore? Would be good to detect that. Maybe a warning would make sense here instead of in the normal case as @lavalamp commented correctly.